### ISSUE TITLE
Reconsider using TypeScript

### ISSUE THREAD

**Issue by paulrobertlloyd** [hirusi/Celestial#76]
Building this app using TypeScript introduces an unnecessary barrier to entry for those wishing to contribute to this project or adapt it to their own needs. 

This is especially relevant given these two project goals:

> * Make IndieWeb accessible to less technically inclined folks.
>
> * To make something useful for myself and others while I figure out and learn backend content.

TypeScript may be fashionable, but it’s worth remembering that it comes with a cost. As far as I’m aware, it was primarily designed to solve the problem of building large applications within large teams, neither of which is the case with this project.

**Comment by hirusi** [comment]
All valid points.

The goal of being accessible certainly wasn't thought of as being accessible to developers (who I would argue _are_ technically inclined to various degrees), rather the end-users who should have an easy to use interface for publishing. Perhaps that wasn't _as_ clear in the text I wrote, and my apologies for that.

I do not believe it conflicts with my goal to learn more -- as using TypeScript is in fact a part of that very goal. I'm currently looking for professional opportunities and TypeScript happens to be, whether I like it or not, a part of the mainstream now and very much a real requirement on job listings.

However, all that said, I had, after my experience using TypeScript on this project so far, [blogged in favour of a relatively simpler solution such as JSDoc](https://rusingh.com/articles/2020/08/15/javascript-jsdoc/) for small and even mid-size projects. I also have felt the [fatigue of using TypeScript in a smaller project](https://fosstodon.org/@celia/104847194910107548) so I'm not against opting out of TypeScript even at this stage.

Would you have any pointers/recommendations around the process of moving to vanilla JS?

I'm also curious - do you personally intend to contribute or adapt it to your own needs in its current state (pre-production release)?  So far, I haven't had anyone express their interest in contributions. If there is such interest, it would certainly give me more reasons to think this over and perhaps plan for it.

**Comment by paulrobertlloyd** [comment]
Huzzah! Guess I should subscribe to your blog ☺️ It’s great to see that you’ve already been thinking about this issue.

I ask having pointed my friend to your project, and he was put-off by the use of TypeScript – as I am whenever I look to projects that I may wish to extend or adapt. His response prompted this issue, as I think it is relevant and worth considering. 

This is also a topic that’s come up in discussions with other people in the IndieWeb community developing projects with JavaScript (see https://github.com/grantcodes/future-micropub-endpoint/issues/2).

**Comment by hirusi** [comment]
😊

I've had time to think about and discuss this further. For several reasons -- I'll explain them below -- I'm going to mark this for v2.

- I've looked around and there doesn't appear to be any tooling to _remove_ TypeScript from a project. (Am I stuck with a bad design choice? Probably. But I'm learning and I would not have expected the community to request a JS-only project.) I'm afraid this becomes a task that must be performed by hand. Say, manually writing out JSDoc comments while removing the pertinent TS code for that JSDoc block. I'm sure one could automate it but I'd rather just do it by hand and get it over with, than frustrate myself with making a utility covering all edge cases.
- Yes, the project isn't so big for the previous point to be a "big issue." However, as I take [a break from Celestial and FOSS](https://fosstodon.org/web/statuses/104874373158408705) in general in [search of gainful employment](https://fosstodon.org/web/statuses/104874251861357806), I wouldn't want to sink time into removing something while we've not reached a v1. I expect this to take several more months, not because of TS slowing me down, but because I'll be spending less time on it.
- Speaking of v1, I think we're all guilty, or have been at some point, of starting projects that never finish. That's not the fate I want for Celestial. Since I dog-food it, I imagine I won't give up on it. Still, my priority is putting out a stable v1 more than anything else.
- With gainful employment, my time to this project will likely reduce to once a week. Context loss is real in such a scenario (at least in my experience and _for me_), and TypeScript helps me make sense of the complex back-end logic so I can return to work quicker.

However, as I mentioned previously, I do want to opt out of TS at some stage, and I think v2 is a more realistic milestone. You're right -- TypeScript _is_ fashionable to an extent, especially for small projects. As was coffeescript, and nobody uses it anymore! I think TS will have a much greater shelf life as it solves a real issue rather than just being syntactically friendly (which was not my experience with coffeescript). It is with this expectation that I also feel comfortable with putting this for v2. TS is going to be around long enough for me to feel at ease opting out at a later date.

I am exploring a JavaScript-only solution but with type-checking that throw errors. My project mentor in fact pointed me to [this talk by Godban Plieshka](https://www.youtube.com/watch?v=kB0G9nCaCSE) which seems like a good solution and very likely what I'll go with.

**Comment by paulrobertlloyd** [comment]
@hirusi All sounds good to me. Thanks for considering my request ☺️

**Comment by hirusi** [comment]
Absolutely! It's a good request and helped me get my thoughts in order. 😸
